src/toil/job.py

```diff
@@ -59,246 +59,897 @@
-        Job._parseResource('memory', '1G')
+        >>> Requirer._parseResource('cores', '1G'), Requirer._parseResource('disk', '1G'), \
+        Requirer._parseResource('memory', '1G')
         (1073741824, 1073741824, 1073741824)
-        >>> Job._parseResource('cores', 1.1)
+        >>> Requirer._parseResource('cores', 1.1)
         1.1
-        >>> Job._parseResource('disk', 1.1) # doctest: +IGNORE_EXCEPTION_DETAIL
+        >>> Requirer._parseResource('disk', 1.1) # doctest: +IGNORE_EXCEPTION_DETAIL
         Traceback (most recent call last):
         ...
         TypeError: The 'disk' requirement does not accept values that are of <type 'float'>
-        >>> Job._parseResource('memory', object()) # doctest: +IGNORE_EXCEPTION_DETAIL
+        >>> Requirer._parseResource('memory', object()) # doctest: +IGNORE_EXCEPTION_DETAIL
         Traceback (most recent call last):
         ...
         TypeError: The 'memory' requirement does not accept values that are of ...
         """
-        assert name in ('memory', 'disk', 'cores')
+        
         if value is None:
+            # Anything can be None.
             return value
-        elif isinstance(value, (str, bytes)):
-            value = human2bytes(value)
-        if isinstance(value, int):
+        
+        if name in ('memory', 'disk', 'cores'):
+            # These should be numbers that accept things like "5G".
+            if isinstance(value, (str, bytes)):
+                value = human2bytes(value)
+            if isinstance(value, int):
+                return value
+            elif isinstance(value, float) and name == 'cores':
+                # But only cores can be fractional.
+                return value
+            else:
+                raise TypeError(f"The '{name}' requirement does not accept values that are of type {type(value)}")
+        elif name == 'preemptable':
+            if isinstance(value, str):
+                if value.tolower() == 'true':
+                    return True
+                elif value.tolower() == 'false':
+                    return False
+                else:
+                    raise ValueError(f"The '{name}' requirement, as a string, must be 'true' or 'false' but is {value}")
+            elif isinstance(value, int):
+                if value == 1:
+                    return True
+                if value == 0:
+                    return False
+                else:
+                    raise ValueError(f"The '{name}' requirement, asn an int, must be 1 or 0 but is {value}")
+            elif isinstance(value, bool):
+                return value
+            else:
+                raise TypeError(f"The '{name}' requirement does not accept values that are of type {type(value)}")
+        else:
+            # Anything else we just pass along without opinons
             return value
-        elif isinstance(value, float) and name == 'cores':
+        
+    def _fetchRequirement(self, requirement):
```

Comment:
What do you mean by "typing"? Documentation that `requirement` is `str` and the return type can be `int`, `float`, or `bool`, depending on the particular requirement queried and its value?

Reply:
Specifically: `def _fetchRequirement(self, requirement: str):`

See PEP484: https://www.python.org/dev/peps/pep-0484/#type-definition-syntax

Also, I just personally prefer this form of type hinting, which is probably bias I admit, but I'd like to start using it where possible in Toil going forward.  Type hints in docstrings are also pretty nice, but I still want to encourage PEP484 type hints.

Reply:
Converting all our docstring types to PEP484 might be A Project, but I've never really fully understood the docstring type syntax anyway, because I don't use an editor that speaks it. PEP484 looks like it actually comes with tooling that checks everything.

We should probably change our contributing doc and open an issue to move over. Maybe there's a tool that can convert the properly-formatted docstring type annotations?